keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+  the one possible without numeric parameters. Ideally, users should not notice
+  at all that a driver is using numeric parameters under the hood.
+
+- Support adding and versioning multiple different numeric models without
+  having to modify the implementation of DRA or the numeric parameters
+  extension.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support numeric class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Define exactly one numeric model that covers all current and future resources.
+  There cannot be a "one size fits all" solution.
+
+## Proposal
+
+### Publishing node capacity
+
+A numeric model defines a capacity type. The type has its own API group, kind
+and potentially different versions. It must support encoding and decoding JSON
+with an apimachinery scheme. The kubelet queries the registered DRA driver
+kubelet plugins for capacity information through a new gRPC call. Drivers
+respond by returning the JSON encoding of this capacity type. A streaming gRPC
+API allows drivers to update the capacity information.
+
+For example, a simple counter could be described as:
+```
+const GroupName = "counter.dra.config.k8s.io"
+const KindName = "Capacity"
+const Version = "v1beta1"
+
+type Capacity struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta `json:",inline"`
+
+	Count int64 `json:"count"`
+}
+```
+
+The kubelet then combines all of this information in a new NodeResourceCapacity
+object which contains the JSON representation of a capacity type in a
+`runtime.RawExtension`. The kubelet itself never has to interpret that data.
+
+### Using numeric parameters as claim parameters
+
+The numeric model also defines a type for claim parameters. In the simple example
+above this is the same `Capacity` type, but more complex models will have
+a more complex capacity type and a simpler parameter type.
+
+When defining the CRD for the claim parameters, the driver developer replicates
+those fields from the numeric model which are relevant for the driver and adds
+custom validation. For a complex model, this can be a true subset of the fields
+in the parameter type. The validation can set bounds that are specific to the
+driver. One can think of this as "duck typing": the CRD doesn't need to include
+the numeric type, a subset of it just needs to behave like one.
+
+To enable extracting the numeric parameters, they must be located at a specific
+JSON path (e.g. "spec"). "Kind" and "apiVersion" must be included at that
+path. Users don't need to set those, the CRD can add them via defaults. This is
```

Comment:
An alternative for this "JSON path + 1:1 mapping of fields" is to store a Go template in the class. Expanding that template must result in a YAML or JSON representation of the numeric parameter type. The data for the [Execute](https://pkg.go.dev/text/template#Template.Execute) call will be the `unstructured.Unstructured` representation of the parameter object.

This template here should match the example:
```
kind: Capacity
apiVersion: counter.dra.config.k8s.io/v1alpha1
count: {{.spec.count}}
```

This removes the need to embed a custom `TypeMeta` inside the CRD.

The downside is the loss of the strong correlation between CRD version and numeric parameter type version: this template will be used for *all* CRD versions and always produces `counter.dra.config.k8s.io/v1alpha1`.

This can be fixed, it just gets more complicated:
```
{{if eq .apiVersion "dra.e2e.example.com/v1alpha1"}}
kind: Capacity
apiVersion: counter.dra.config.k8s.io/v1alpha1
count: {{.spec.count}}
{{end}}
```

If the `ClaimParameter` type ever gets versioned, then additional if blocks need to be added for the new versions.

I'm not sure whether I got the Go template syntax right. We'll probably have to add some test tools for this or a Go playground example.

Reply:
I'm not a fan of Go templates either, so if we have something better then I am open to that.

> No loops but we could add functions that for example take an array and find the min/max etc.

We need to be careful about what we allow: the existence of such functions and their parameters become part of the Kubernetes API. Do we perhaps a stable set of functions that we can refer to instead of defining (and implementing) our own set?

Go templates have the same problem, but at least those come with some "batteries included".

Reply:
So does CEL in Kubernetes, and there is already some strong versioning support:
https://kubernetes.io/docs/reference/using-api/cel/

I've [prototyped templating](https://github.com/kubernetes/kubernetes/pull/123033), but must be missing something because I get an error when parsing expressions. I asked for help and will keep looking...